### PR TITLE
Feature/like songs/like locally

### LINKED ISSUE #309: Have a like counter on the track profile association

### PR DESCRIPTION
**Pull request by IB-12**
This allows users to like (and unlike) songs in beacons, which adds them to a dedicated list and increases the song's OP's like counter.

Closes #309

### CODE REVIEW

File: app/src/main/java/ch/epfl/cs311/wanderwave/model/data/ProfileTrackAssociation.kt [hunk 1]
```diff
@@ -9,9 +9,24 @@ package ch.epfl.cs311.wanderwave.model.data
...
+    val likers: List<Profile> = emptyList(),
+    val likes: Int = 0
+) {
 
   fun toMap(): Map<String, Any?> {
     return hashMapOf("profile" to profile?.toMap(), "track" to track.toMap())
   }
+
+  fun likeTrack(profile: Profile): ProfileTrackAssociation {
+    this.profile?.numberOfLikes = this.profile?.numberOfLikes!! + 1
```
Comment (KlaKalma) [inline app/src/main/java/ch/epfl/cs311/wanderwave/model/data/ProfileTrackAssociation.kt:24]: This seems like a bad null assert, Can you add a test before that test if the profile is not null, 
something like 
`this.profile?.let{...}
?: // case if it's null`

Comment (KlaKalma) [reply app/src/main/java/ch/epfl/cs311/wanderwave/model/data/ProfileTrackAssociation.kt:24]: And maybe add a safety check that the profile is not already in the likers.

File: app/src/main/java/ch/epfl/cs311/wanderwave/model/data/ProfileTrackAssociation.kt [hunk 2]
```diff
@@ -9,9 +9,24 @@ package ch.epfl.cs311.wanderwave.model.data
...
+    return ProfileTrackAssociation(profile, track, likers + profile, likes + 1)
+  }
+
+  fun unlikeTrack(profile: Profile): ProfileTrackAssociation {
+    this.profile?.numberOfLikes = this.profile?.numberOfLikes!! - 1
```
Comment (KlaKalma) [inline app/src/main/java/ch/epfl/cs311/wanderwave/model/data/ProfileTrackAssociation.kt:29]: Same here, null assert and safety check

File: app/src/main/java/ch/epfl/cs311/wanderwave/viewmodel/ProfileViewModel.kt [hunk 3]
```diff
@@ -229,6 +232,16 @@ constructor(
     _childrenPlaylistTrackList.value = (emptyList())
   }
 
+  fun likeTrack(track: Track) {
+    _wanderwaveLikedTracks.value += track
+    Log.d("ProfileViewModel", "Liked track: $track  " + _wanderwaveLikedTracks.value)
```
Comment (KlaKalma) [inline app/src/main/java/ch/epfl/cs311/wanderwave/viewmodel/ProfileViewModel.kt:237]: Log

File: app/src/main/java/ch/epfl/cs311/wanderwave/viewmodel/ProfileViewModel.kt [hunk 4]
```diff
@@ -229,6 +232,16 @@ constructor(
...
+  }
+
+  fun unlikeTrack(track: Track) {
+    _wanderwaveLikedTracks.value -= track
+    Log.d("ProfileViewModel", "Uniked track: $track  " + _wanderwaveLikedTracks.value)
```
Comment (KlaKalma) [inline app/src/main/java/ch/epfl/cs311/wanderwave/viewmodel/ProfileViewModel.kt:242]: Log

**Review by KlaKalma — CHANGES_REQUESTED** [review]
Just some little comments nothing very bothering, I like the new like button

**Comment by IB-12** [comment]
Thanks for the review, your inputs are greatly appreciated

File: app/src/main/java/ch/epfl/cs311/wanderwave/model/data/ProfileTrackAssociation.kt [hunk 5]
```diff
@@ -20,13 +20,23 @@ data class ProfileTrackAssociation(
...
     return ProfileTrackAssociation(profile, track, likers + profile, likes + 1)
   }
 
   fun unlikeTrack(profile: Profile): ProfileTrackAssociation {
-    this.profile?.numberOfLikes = this.profile?.numberOfLikes!! - 1
+    if (!likers.contains(profile)) return this
+    if (this.profile != null) {
+      this.profile.numberOfLikes -= 1
+    }
     return ProfileTrackAssociation(profile, track, likers - profile, likes - 1)
```
Comment (KlaKalma) [inline app/src/main/java/ch/epfl/cs311/wanderwave/model/data/ProfileTrackAssociation.kt:40]: Sorry for bothering, maybe I'm missing something but wouldn't this return need to be in the if? Because if the profile already like this song we don't want to add it again to the likers no?

**Review by KlaKalma — CHANGES_REQUESTED** [review]

File: app/src/main/java/ch/epfl/cs311/wanderwave/model/data/ProfileTrackAssociation.kt (hunk 5 above)
Comment (IB-12) [reply app/src/main/java/ch/epfl/cs311/wanderwave/model/data/ProfileTrackAssociation.kt:40]: Yeah that's why there is `if (!likers.contains(profile)) return this` just above, which returns an unchanged ProfileTrackAssociation if the profile has already liked the song

Comment (KlaKalma) [reply app/src/main/java/ch/epfl/cs311/wanderwave/model/data/ProfileTrackAssociation.kt:40]: I meant adding null to the list, if profile is null we don't want to add it to the list no?

**Comment by KlaKalma** [comment]
Perfect sir Thanks !

**Review by KlaKalma — APPROVED** [review]
lgtm
